### ISSUE TITLE
Identifying the Property from DerivedProperty/Multilink

### ISSUE THREAD

**Issue by zepumph** [phetsims/axon#259]
Very similar to https://github.com/phetsims/axon/issues/187, but this time for listeners to DerivedProperty and Multilink. In separate meetings with @twant and @jessegreenberg today I ran into two cases where it would be nice to know what Property triggered a listener. Has anyone ran into this before?

Both of the cases have to do with description implementation. Basically it boils down to how we need more context about what changed in the sim, and the Property that changes influences what the description update is.

* For @twant and I, in Molarity we have two sliders that have almost identical conditional logic for determining the alert text of each slider. The only thing that differs is the value of the string. See `MolarityAlertManager` linkage to `volumeProperty` and `soluteAmountProperty`. 

* For @jessegreenberg and I, in Gravity Force Lab Basics both the constant radius checkbox and the mass controls effect the `radiusProperty` of a puller, but the alert differs depending on which one changes the radius. 

Is this something that anyone else is interested in having?

I am thinking it could be as easy as tacking the instance of the changed Property to the end of the argument list. Something like:

```js
super.set( derivation.apply( null, dependencies.map( property => property.get() ).push( changedProperty ) );
```

-----------

It would also be good to note that it is very likely this will come up again for description work. 

Looking forward to discussion!

**Comment by pixelzoom** [comment]
Instead of making this the responsibility of the listener for the DerivedProperty or Multilink (which should not care which Property changed), could you add a listener directly to the Property that you are interested in?

**Comment by samreid** [comment]
Perhaps we should move toward a paradigm where user interface components also trigger listeners (along with changing a `Property`).  For molarity, that would mean the alert text is set by the volume property slider or the solute amount property slider.  For Gravity Force Lab Basics, there would be a listener on the constant radius checkbox and another on the mass controls.  I don't know if it's practical to apply this philosophy to our entire stack, but it seems like it relates somewhat to the sonification problem of having to mute sounds during reset all--if sounds were wired up to user interface components instead of model properties, that problem would go away.

Also, this reminds me of a pattern @chrisklus and I explored in https://github.com/phetsims/phet-io/issues/1349#issuecomment-422221687.  We prototyped a new component UserControlledProperty which was summarized like so:

> To accomplish this, we introduced a new type UserControlledProperty that has the same API as Property, but which tracks whether the UI set the value. This change does not require any changes to sun components. Full details are in the two branches noted above. It is unclear whether this pattern is worth the trouble.

**Comment by jonathanolson** [comment]
One workaround is to add independent listeners to each "source" Property, and turn the DerivedProperty into a normal read-only Property that is modified by each listener.

We've also used boolean flag-based systems in the past (like DynamicProperty) where it tracks the source of the change. Additionally https://github.com/phetsims/axon/issues/251 is related, where it may be beneficial to know when a Property is notifying its listeners. It's also possible to use a listener pattern like @samreid mentioned above, and that could be valuable.

This is also closely related to sonification, where it's necessary to know whether a change in a Property has come from a sim reset (and we could have multiple other causes/states we may need to disambiguate).

**Comment by zepumph** [comment]
>We've also used boolean flag-based systems in the past (like DynamicProperty) where it tracks the source of the change.

In https://github.com/phetsims/gravity-force-lab-basics/issues/168 I used flags like this to see what had been updated since the last step, and this was able to give the context necessary. I think that in the future, I will likely use the view driven approach of adding listeners to UI components instead of model Properties when possible. Thanks all for the advice. I think we should add these strategies to documentation somewhere, though I'm not sure what md it belongs in. It isn't quite a design pattern, but maybe it could be a subsection in the "Observer" pattern.
